src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -747,6 +748,66 @@ private bool CheckFieldValueKind(SyntaxNode node, BoundFieldAccess fieldAccess,
             return CheckIsValidReceiverForVariable(node, fieldAccess.ReceiverOpt, valueKind, diagnostics);
         }
 
+        bool IsAllowedInitOnlyAssignment(Symbol accessedMember, BoundExpression receiver, bool restrictToThis)
+        {
+            // ok: new C() { InitOnlyProperty = ... }
+            if (receiver is BoundObjectOrCollectionValuePlaceholder)
+            {
+                return true;
+            }
+
+            // bad: other.InitOnlyProperty = ...
+            if (!(receiver is BoundThisReference) && !(receiver is BoundBaseReference) && !(receiver is BoundObjectOrCollectionValuePlaceholder))
+            {
+                return false;
+            }
+
+            var containingMember = ContainingMemberOrLambda;
+            if (!(containingMember is MethodSymbol method))
+            {
+                return false;
+            }
+
+            if (!isDerivedType(method.ContainingType, accessedMember.ContainingType.OriginalDefinition, restrictToThis))
+            {
+                return false;
+            }
+
+            if (method.IsConstructor() || isInitOnlySetter(method))
+            {
+                // ok: setting on `this` or `base` from a constructor or init-only setter
```

Comment:
>// ok: setting on `this` or `base` from a constructor or init-only setter [](start = 16, length = 73)

I believe we decided disallow assigning to readonly fields from base, a class can assign readonly fields that it itsef declares and no other readonly fields.

Reply:
> I believe we decided to disallow assigning to readonly fields from base, a class can assign readonly fields that it itsef declares and no other readonly fields.

This was resolved by design with no explanation. However, the speclet says: "The ability to assign readonly fields from an init accessor is limited to those fields declared on the same type as the accessor. It cannot be used to assign readonly fields in a base type." Etc.

---
In reply to: [411009843](https://github.com/dotnet/roslyn/pull/43275#discussion_r411009843) [](ancestors = 411009843)

Reply:
Unless I'm misunderstanding the scenario you have in mind, the implementation is compliant with the spec. A test to illustrate this is `ReadonlyFields`.

---
In reply to: [413399060](https://github.com/dotnet/roslyn/pull/43275#discussion_r413399060) [](ancestors = 413399060,411009843)

Reply:
> Unless I'm misunderstanding the scenario you have in mind, the implementation is compliant with the spec. A test to illustrate this is ReadonlyFields.

The comment was for iteration 2. I believe the code was reachable for readonly fields and looked like this (specifically note the comment):
```
            if (method.IsConstructor() || isInitOnlySetter(method))
            {
                // ok: setting on `this` or `base` from a constructor or init-only setter
                return true;
            }
```

If you fixed it, that is one thing. Otherwise I would like to understand what was there by design. Just want to make sure we are on the same page.

---
In reply to: [413407114](https://github.com/dotnet/roslyn/pull/43275#discussion_r413407114) [](ancestors = 413407114,413399060,411009843)

Reply:
I should have been clearer, your expectation was correct, implemented and verified.

---
In reply to: [413412567](https://github.com/dotnet/roslyn/pull/43275#discussion_r413412567) [](ancestors = 413412567,413407114,413399060,411009843)